### PR TITLE
ENH: Add OpenSSF Scorecard GitHub Action

### LINKED ISSUE #22476: ENH: Add the OpenSSF Scorecards GitHub Action
### Proposed new feature or change:

Hi, I'm Pedro and I'm working for Google and the [Open Source Security Foundation][ossf] to help essential open-source projects improve their supply-chain security. Given that NumPy is, to quote the moto, "the fundamental package for scientific computing with Python," the OpenSSF has identified it as one of the 100 most critical open source projects.

NumPy has already collaborated with the OpenSSF in #20584 (shipping 2FA tokens to NumPy maintainers). Continuing our efforts to improve the security of open-source software, I'd like to suggest the adoption of a new OpenSSF tool, the [Scorecards GitHub Action][sc-gha].

The [Scorecards][sc] system runs dozens of automated security [checks][checks] to help maintainers better understand their project's supply-chain security posture. It is developed by the OpenSSF, [in partnership with GitHub][sc-blog].

The Action performs these checks automatically after every commit to the main branch. The results of its checks are available on the project's [security dashboard](https://github.com/numpy/numpy/security), and include suggestions on how to solve any issues (see an example below). It's already been adopted by 1800+ projects, including [Tensorflow][tensorflow],  [PyTorch][pytorch], [Angular][angular], and [Flutter][flutter].

Would you be interested in a PR which adds this Action? Optionally, it can also publish your results to the OpenSSF REST API, which allows a [badge][badge] with the project's score to be added to its README.

If you have any questions about Scorecards, you can check out its [FAQ][faq] or just ask me!

![Detail of a Token-Permissions alert, indicating the specific file and remediation steps][img-detail]

[angular]: https://github.com/angular/angular  
[badge]: https://openssf.org/blog/2022/09/08/show-off-your-security-score-announcing-scorecards-badges/
[checks]: https://github.com/ossf/scorecard#scorecard-checks
[faq]: https://github.com/ossf/scorecard/blob/main/docs/faq.md#frequently-asked-questions
[flutter]: https://github.com/flutter/flutter
[ossf]: https://openssf.org/
[pytorch]: https://github.com/pytorch/pytorch
[sc]: https://github.com/ossf/scorecard
[sc-blog]: https://github.blog/2022-01-19-reducing-security-risk-oss-actions-opensff-scorecards-v4/
[sc-gha]: https://github.com/ossf/scorecard-action
[tensorflow]: https://github.com/tensorflow/tensorflow

[img-detail]: https://user-images.githubusercontent.com/15221358/190184600-ee8d3b39-077e-416a-8711-1b5fb01cf0b3.png

### PR DESCRIPTION
**Pull request by pnacht**
Closes #22476 

As per the linked issue, this PR adds the OpenSSF Scorecard GitHub Action.

I have taken the liberty of adding the badge with the project's score to its README. It's meant to let the project's consumers have an at-a-glance understanding of the project's security posture. However, the badge is entirely optional, just say the word and I'll quickly remove it.

### CODE REVIEW

**Review by seberg — APPROVED** [review]
I have not explored this, but it looks like a good start.  Personally, I am not sure I like the badge much, but unlike testing batches it should at least not show failures randomly.

Only checking `main` for protection rules seems a bit strange to me, since releases are made of the other branches after all?

**Comment by pnacht** [comment]
@seberg Ah, to clarify: the comment regarding branch protection rules is actually a [GitHub implementation detail](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#branch_protection_rule):

> Note: This event will only trigger a workflow run if the workflow file is on the default branch.

The Scorecards workflow checks all development and release branches for branch protection rules.

As for the badge, let me reiterate that it is entirely optional. It's offered as a way for the project's consumers to have an at-a-glance understanding of the project's security posture. But the choice to adopt it is entirely yours. Say the word and I'll remove it.

As for the failing builds, I'm not sure what to do. The Travis runs failed with:

> Automatic restarts limited: Please try restarting this job later or contact support@travis-ci.com.

and the Azure runs seem to fail on PyTests.

**Comment by seberg** [comment]
Don't worry about the failures, it is just flaky (e.g. new cygwin and pytest right now also).

Understood about the badge, just thought I would give someone with a second opinion a chance to decide either way and then merge.

**Comment by rgommers** [comment]
Thanks @pnacht. I'll have a more detailed look in the coming days. 

For other reviewers, see https://api.securityscorecards.dev/projects/github.com/numpy/numpy for current scores/checks.

The badge gives a number, `6.2` in our case. I'm not sure many people know how to interpret that number - it feels like a low score, but it's probably because some of the scored items aren't applicable. I plan to dig into that a bit more:

image

**Comment by mattip** [comment]
I wonder how they weight the scores. Here are the things they really didn't like:
- "non read-only tokens detected in GitHub workflows" (see #22481)
- "dependency not pinned by hash detected -- score normalized to 0"
- "SAST tool is not run on all commits -- score normalized to 0" (that sounds like a bit of self-promotion...)
- "Determines if the project has an OpenSSF (formerly CII) Best Practices Badge."
- "0 out of 5 artifacts are signed or have provenance" (based on the github release artifacts, without considering PyPI)

and we score 3 on
- "5 out of last 16 changesets reviewed before merge -- score normalized to 3"
- "branch protection is not maximal on development and all release branches"

**Comment by charris** [comment]
> "5 out of last 16 changesets reviewed before merge

I guess that means we need to actually hit the review button when doing reviews. If only there were a "none of the above" option :)

**Comment by charris** [comment]
> "0 out of 5 artifacts are signed or have provenance"

Can we turn off Github's automatic generation of bad artifacts? That would reduce that to 0 out of 3. We do publish the hash for the source files, what do they want there?

**Comment by rgommers** [comment]
> Can we turn off Github's automatic generation of bad artifacts?

Last time I checked that was not possible.



> We do publish the hash for the source files, what do they want there?

I'll note that some of these questions take time to dig into - I plan to do that as part of a Tidelift-supported effort, where they will also evaluate which metrics actually make sense. It's already clear that that's not the case for all of them (e.g., pinning all dependencies makes sense for applications, not for libraries).

**Comment by pnacht** [comment]
> The badge gives a number, `6.2` in our case. I'm not sure many people know how to interpret that number - it feels like a low score, but it's probably because some of the scored items aren't applicable. I plan to dig into that a bit more:

@rgommers 6.2 is actually a pretty good score, putting NumPy at around the top 30% of very important projects* (or top 3% of a much broader sample of GitHub projects overall). Your reaction that it's a low score is a feedback we've heard before and something we're actively working to improve!

> I wonder how they weight the scores.

@mattip The final score (6.2) is calculated via a weighted average of all the component scores (see [here](https://securityscorecards.dev/#how-it-works) for the weights and [here](https://github.com/ossf/scorecard/blob/main/docs/checks.md) for more detailed explanations of each check).

> * "SAST tool is not run on all commits -- score normalized to 0" (that sounds like a bit of self-promotion...)

Just to clarify, SAST is simply an acronym for "static application security testing", more commonly known as "static analysis". Scorecards currently recognizes [CodeQL](https://codeql.github.com/), [LGTM](https://lgtm.com/) and [SonarCloud](https://sonarcloud.io/) as SAST tools. CodeQL and LGTM are GitHub proprietary tools, SonarCloud is developed by Sonar.

> * "Determines if the project has an OpenSSF (formerly CII) Best Practices Badge."

The [OpenSSF Best Practices Badge](https://bestpractices.coreinfrastructure.org/en) is a self-reported questionnaire on the project's security posture. It's presence is included in the score as a sort of heuristic for whether the project uses security best practices. It's questions are more granular and cover information that is not available via the GitHub API. 

> * "5 out of last 16 changesets reviewed before merge -- score normalized to 3"
> * "branch protection is not maximal on development and all release branches"

> I guess that means we need to actually hit the review button when doing reviews. If only there were a "none of the above" option :)

@charris Actually not quite! The Code-Review check checks whether a commit came via PR and whether that PR went through code review. However, it accepts two types of code-reviews: explicit and implicit:

* Explicit are those where a reviewer is determined, gives their approval and the PR is merged.
* Implicit is any time a PR is merged by someone other than the author. That is, Scorecards trusts that whoever merged the PR looked at the changes first.

The issue here seems to be that many commits skip PRs entirely, instead being pushed directly to main. From what I can see in the history, these tend to be non-code commits (MAINT, TST, DOC, STY etc).

I can entirely understand why such commits are allowed. Scorecards is merely pointing out the risk of an account being compromised and pushing a malicious commit under the guise of a `STY` commit no one bothers to double-check. Whether the security benefits of moving such commits to PRs are worth the overhead costs (i.e. the precious resources that are maintainers' time and patience) is a decision you are all better-suited to make than me!

 > We do publish the hash for the source files, what do they want there?

The checksums offered with the NumPy release are very useful in that they allow users to ensure the files they downloaded are what they expect (there was no corruption or man-in-the-middle attack). However, they can't help users detect a malicious release.

Signed releases basically secure the entire transaction. If a release is cryptographically signed (usually with GPG), users can be sure that it was created by the actual maintainer (or at least someone with access to their private key...).

---

\* arbitrarily defined by yours truly as projects with a [criticality_score](https://github.com/ossf/criticality_score) > 0.5

**Comment by mattip** [comment]
> Just to clarify, SAST is simply an acronym ...

Ahh, thanks. The comment was snarky, sorry. We have had some useful experience when using static analysis on the repo, but it seems like something that should be done manually since so far the tools have had many false positives. Is there a way to declare the last time a SAST was run on the repo?

> The issue here seems to be that many commits skip PRs entirely, instead being pushed directly to main

Weird. The commit 7776bc616 seems to be made directly to main, but comes from #21853. Is that because it was squash-merged?

**Comment by rgommers** [comment]
> The issue here seems to be that many commits skip PRs entirely, instead being pushed directly to main. From what I can see in the history, these tend to be non-code commits (MAINT, TST, DOC, STY etc).

That sounds wrong. Are you sure that those are not PRs that were squash-merged? AFAIK no one is pushing directly to `main` - and we have branch protection enabled, so I think only repo owners can push to `main` to begin with.

**Comment by charris** [comment]
> The issue here seems to be that many commits skip PRs

For releases I push an update directly to the release branch right after pushing the release tag, you shouldn't see any direct pushes in the main branch. The release manager basically owns the release branch, also merges backports without review, because they have already been reviewed.

**Comment by pnacht** [comment]
Sorry for the misunderstanding, everyone! I looked at the Commit History and saw a bunch of commits that weren't "Merged pull request #..." and figured they'd bypassed the PR process (I've grown used to seeing PRs get squash-merged...).

I'll investigate further why so few PRs are getting correctly identified as code-reviewed.

**Comment by seberg** [comment]
We don't have a habit of using the "accepted review" feature on NumPy (nor does numpy currenlty require it, I guess partly also because we are rather few active devs).  I could image that you are counting "review" as that?

**Review by rgommers — APPROVED** [review]
I've now looked at this in more detail, verified the pinned commit hashes of the used actions, and reviewed the OpenSSF Scorecards themselves in more detail. All looks good.

I'll also note that, in addition to the bit of content at https://github.com/numpy/numpy/security, there's a detailed scorecard for NumPy at https://deps.dev/pypi/numpy/1.23.4.

This job will only run when it's merged into `main`. I believe there's one check (branch protection) for which we need to add a token, see https://github.com/ossf/scorecard-action#installation. But let's find out by actually running it.

Let's keep the badge, since it doesn't hurt, the linked data is useful, Tidelift requested it as well, and it may actually increase our score.

In it goes, thanks @pnacht!

**Comment by rgommers** [comment]
Here is the first run of the job (succeeded in 1m 4s): https://github.com/numpy/numpy/actions/runs/3396721101/jobs/5648096177. The [summary](https://github.com/numpy/numpy/actions/runs/3396721101) of that job has a `results.sarif` artifact attached to it, with the detailed results. That will take some time to go through; there are things we can improve over time.

**Comment by rgommers** [comment]
Yep, branch protection check needs the token:

```
name | "Branch-Protection"
-- | --
score | -1
reason | "internal error: error during branchesHandler.setup: internal error: githubv4.Query: Resource not accessible by integration"
 ```

@pnacht our score improved to `7.4`, I did not expect that. May be because of more recent results, but perhaps there's a difference in calculation between the action and however the number is calculated elsewhere?

**Comment by pnacht** [comment]
Looking at the previous score (unfortunately currently requires using the [public BigQuery data](https://github.com/ossf/scorecard#public-data)), the only significant change was in the Token-Permissions check, which went from 0 to 10. Since it's a "High" risk check, it has a significant weight in the final score.

(And the previous score was generated with a token and was therefore able to make a basic Branch-Protection check, so it considered the `3/10` while your `-1` is ignored).

 I can also tell you that 7.4 puts NumPy at the top 6% of critical projects (and top 0.3% of projects overall).

High fives all around!

**Comment by rgommers** [comment]
> Looking at the previous score (unfortunately currently requires using the [public BigQuery data](https://github.com/ossf/scorecard#public-data)), the only significant change was in the Token-Permissions check, which went from 0 to 10. Since it's a "High" risk check, it has a significant weight in the final score.

Thanks, that is helpful. I will try to figure out what token we changed. I know I added a few recently that I did scope much more tightly.

**Comment by rgommers** [comment]
From https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions:

_The highest score is awarded when the permissions definitions in each workflow's yaml file are set as read-only at the [top level](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions) and the required write permissions are declared at the [run-level](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idpermissions). One point is reduced from the score if all jobs have their permissions defined but the top level permissions are not defined. This configuration is secure, but there is a chance that when a new job is added to the workflow, its job permissions could be left undefined because of human error._

This was probably:
- gh-22367
- gh-22481

**Pull request merged**
